pkg/mcs/resource_manager/client/client.go

```diff
@@ -0,0 +1,796 @@
+		if request != nil {
+			requests = append(requests, request)
+		}
+		return true
+	})
+	if len(requests) > 0 {
+		c.sendTokenBucketRequests(ctx, requests, source)
+	}
+}
+
+func (c *ResourceGroupsController) sendTokenBucketRequests(ctx context.Context, requests []*rmpb.TokenBucketRequest, source string) {
+	now := time.Now()
+	c.run.lastRequestTime = now
+	c.run.requestInProgress = true
+	req := &rmpb.TokenBucketsRequest{
+		Requests:              requests,
+		TargetRequestPeriodMs: uint64(c.config.targetPeriod / time.Millisecond),
+	}
+	go func() {
+		log.Debug("[resource group controllor] send token bucket request", zap.Time("now", now), zap.Any("req", req.Requests), zap.String("source", source))
+		resp, err := c.provider.AcquireTokenBuckets(ctx, req)
+		if err != nil {
+			// Don't log any errors caused by the stopper canceling the context.
+			if !errors.ErrorEqual(err, context.Canceled) {
+				log.L().Sugar().Infof("TokenBucket RPC error: %v", err)
+			}
+			resp = nil
+		}
+		log.Debug("[resource group controllor] token bucket response", zap.Time("now", time.Now()), zap.Any("resp", resp), zap.String("source", source), zap.Duration("latency", time.Since(now)))
+		c.tokenResponseChan <- resp
+	}()
+}
+
+func (c *ResourceGroupsController) handleTokenBucketTrickEvent(ctx context.Context) {
+	c.groupsController.Range(func(name, value any) bool {
+		gc := value.(*groupCostController)
+		gc.handleTokenBucketTrickEvent(ctx)
+		return true
+	})
+}
+
+func (c *ResourceGroupsController) mainLoop(ctx context.Context) {
+	interval := c.config.groupLoopUpdateInterval
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	c.updateRunState(ctx)
+	c.collectTokenBucketRequests(ctx, "init", false /* select all */)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case resp := <-c.tokenResponseChan:
+			c.run.requestInProgress = false
+			if resp != nil {
+				c.updateRunState(ctx)
+				c.handleTokenBucketResponse(resp)
+			} else {
+				// A nil response indicates a failure (which would have been logged).
+				c.run.requestNeedsRetry = true
+			}
+		case <-ticker.C:
+			c.updateRunState(ctx)
+			c.updateAvgRequestResourcePerSec()
+			if c.run.requestNeedsRetry || c.shouldReportConsumption() {
+				c.run.requestNeedsRetry = false
+				c.collectTokenBucketRequests(ctx, "report", false /* select all */)
+			}
+		case <-c.lowTokenNotifyChan:
+			c.updateRunState(ctx)
+			c.updateAvgRequestResourcePerSec()
+			if !c.run.requestInProgress {
+				c.collectTokenBucketRequests(ctx, "low_ru", true /* only select low tokens resource group */)
+			}
+		default:
+			c.handleTokenBucketTrickEvent(ctx)
+		}
+	}
+}
+
+// OnRequestWait is used to check whether resource group has enough tokens. It maybe needs wait some time.
+func (c *ResourceGroupsController) OnRequestWait(
+	ctx context.Context, resourceGroupName string, info RequestInfo,
+) (err error) {
+	if _, ok := defaultWhiteList[resourceGroupName]; ok {
```

Comment:
This check could be done by https://github.com/tikv/client-go/pull/648/files#diff-34812587e42268a5d84933be286601fc58fdff17ab336b09ba75455ffe5ed447R96. I prefer keeping it there.